### PR TITLE
Issue #1093 gltf generator

### PR DESCRIPTION
**Pull request by ikehirzel**
This is the fix for https://github.com/stoqs/stoqs/issues/1093 that was done for CSUMB capstone fall 2021

### CODE REVIEW

**Comment by stoqs** [comment]
Thanks for this pull request!

I’ll test it tomorrow and will let you know if I have any questions.

**Comment by MBARIMike** [comment]
Dropping some notes here on building this under CentOS 7, the OS that's used for a STOQS Vagrant development system...

1. Build netcdf-cxx4 from https://github.com/Unidata/netcdf-cxx4
2. Install cmake3: `yum install -y cmake3`
3. There were some changes I needed to make to the code to get it to compile. Here are the diffs:
```diff
[vagrant@localhost stoqsgit]$ git diff
diff --git a/stoqs/contrib/gltf-generator/CMakeLists.txt b/stoqs/contrib/gltf-generator/CMakeLists.txt
index 0cdbfd2..ec66185 100644
--- a/stoqs/contrib/gltf-generator/CMakeLists.txt
+++ b/stoqs/contrib/gltf-generator/CMakeLists.txt
@@ -1,21 +1,23 @@
 cmake_minimum_required(VERSION 3.13)
 project("grd-to-gltf")

+list(APPEND CMAKE_PREFIX_PATH "/home/vagrant/Downloads/netcdf-cxx4/build")
+
 # getting libraries
-find_package(netCDF CONFIG REQUIRED)
+find_package(netCDFCxx CONFIG REQUIRED)

 if (WIN32)
        add_library(netcdf SHARED IMPORTED)
        set_target_properties(netcdf PROPERTIES IMPORTED_IMPLIB "${netCDF_LIB_DIR}/netcdf.lib")
        set(LIBRARIES "netcdf")
 else()
-       set(LIBRARIES ${netCDF_LIBRARIES})
+       set(LIBRARIES "netcdf")
 endif()

 if (MSVC)
        add_compile_options(/W4)
 else()
-       add_compile_options(-Wall -Wextra -pedantic)
+       add_compile_options(-Wall -Wextra -pedantic -std=c++11)
 endif()

 # getting sources & include dirs
diff --git a/stoqs/contrib/gltf-generator/src/stoqs/bathymetry.cpp b/stoqs/contrib/gltf-generator/src/stoqs/bathymetry.cpp
index 64fba8f..66a68d4 100644
--- a/stoqs/contrib/gltf-generator/src/stoqs/bathymetry.cpp
+++ b/stoqs/contrib/gltf-generator/src/stoqs/bathymetry.cpp
@@ -162,8 +162,8 @@ namespace stoqs
                _xysize = _z.count();
                _dimension[0] = _z.size_x();
                _dimension[1] = _z.size_y();
-               _spacing[0] = std::abs(_x_range[1] - _x_range[0]) / (double)(_dimension[0] - 1);
-               _spacing[1] = std::abs(_y_range[1] - _y_range[0]) / (double)(_dimension[1] - 1);
+               _spacing[0] = std::labs(_x_range[1] - _x_range[0]) / (double)(_dimension[0] - 1);
+               _spacing[1] = std::labs(_y_range[1] - _y_range[0]) / (double)(_dimension[1] - 1);
        }

        std::string Bathymetry::to_string() const
diff --git a/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp b/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp
index 4495916..c0e0f25 100644
--- a/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp
+++ b/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp
@@ -1,7 +1,7 @@
 #include <stoqs/geometry.h>

 // standard library
-#include <math.h>
+#include <cmath>

 #define EARTH_RADIUS_M 6371000.0
```
Now that it compiles for me I'll test it on some .grd files!

**Comment by MBARIMike** [comment]
Here's a report on testing with the Monterey25.grd file…

1. I created a .glb file with 10x vertical exaggeration from Monterey25.grd:
```
./grd-to-gltf Monterey25.grd -e 10 -b
```
2. I put the resulting file into a public X3D scene at https://stoqs.mbari.org/x3d/Monterey25_10x/Monterey25_e10.html

It's convenient to have the terrain viewable here as we can use the Debug Log display to verify the coordinates of the terrain:
1. After the 252 MB uncompressed file loads (about 10 seconds) press the "A" key to see all objects in the scene
2. Check the Log box to display X3DOM's log output
3. In Examine mode (the default) navigate to an interesting viewpoint
4. Press the "V" key to print out the position, orientation, and centerOfRotation values

These are the values used for the default viewpoint for a STOQS campaign, e.g. in [this loader](https://github.com/stoqs/stoqs/blob/20ec9d4cb6f2a33d9b8359f15149347cf6526a7a/stoqs/loaders/loadTestData.py#L34-L36) for the default campaign.

File: stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	double Geometry::get_latitude(const Bathymetry& bathymetry, size_t y)
+	{
+		return bathymetry.latitude_min() + bathymetry.latitude_spacing() * (double)y;
+	}
+
+	Vertex Geometry::get_earth_centered_vertex(double longitude, double latitude, double altitude, uint32_t id)
+	{
+		double phi = to_radians(latitude);
```
Comment (MBARIMike) [inline stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp:32]: Comparing this code to the [GMT code](https://github.com/GenericMappingTools/gmt/blob/be890649579be45e94269632786d08890a26dfea/src/gmt_map.c#L9094-L9108) indicates that we probably need to apply the WGS-84 parameters here

**Review by MBARIMike — COMMENTED** [review]
I will change the geometry.cpp code to convert to WGS-84.

**Comment by MBARIMike** [comment]
Some more details on the coordinate conversion...

A WGS-84 coordinate converted to ECEF using gmt(1):
```
echo "-122.503094 36.440848 -1020.000000" | gmt mapproject -E
-2759951.41997	-4331739.72409	3767050.17337
```
With the code here [modified to print out the vertex coordinates](https://gist.github.com/MBARIMike/6f3d7a22963796a028ac9885900d820c) for the same WGS-84 coordinate:
```
./grd-to-gltf Monterey25.grd -e 10 -b
longitude, latitude, altitude: -122.503094 36.440848 -1020.000000
x, y, z: -2753604.250000 -4321778.000000 3783720.750000
``` 
The x, y, z ECEF values are close, but not correct.

**Comment by MBARIMike** [comment]
For reference here is the [X3DOM code that converts WGS-84 to GeoCentric (ECEF)](https://github.com/x3dom/x3dom/blob/3ace18318cd192e424546569932abfe3e1e2346a/src/nodes/Geospatial/GeoCoordinate.js#L380-L443).

**Comment by MBARIMike** [comment]
A Google search for documentation on the conversion math reveals this from an old friend:
https://www.oc.nps.edu/oc2902w/coord/coordcvt.pdf

**Review by MBARIMike — APPROVED** [review]
Tested WGS-84 code and compared with other data. All is good with those changes which will be contributed in a follow on PR.

**Comment by MBARIMike** [comment]
Capturing here the changes made in geometry.cpp to get the resulting .glb to agree with other terrain data:
```diff
[vagrant@localhost stoqsgit]$ git diff stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp
diff --git a/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp b/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp
index 4495916..9ff39fb 100644
--- a/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp
+++ b/stoqs/contrib/gltf-generator/src/stoqs/geometry.cpp
@@ -1,9 +1,12 @@
 #include <stoqs/geometry.h>

 // standard library
-#include <math.h>
+#include <cmath>
+#include <iostream>

 #define EARTH_RADIUS_M 6371000.0
+#define WGS_84_SEMI_MAJOR_AXIS 6378137.0
+#define WGS_84_INVERSE_FLATTENING 298.257223563

 namespace stoqs
 {
@@ -24,11 +27,24 @@ namespace stoqs

        double Geometry::get_latitude(const Bathymetry& bathymetry, size_t y)
        {
-               return bathymetry.latitude_min() + bathymetry.latitude_spacing() * (double)y;
+               //std::cerr << std::fixed;
+               //std::cerr << "bathymetry.latitude_min(): " << bathymetry.latitude_min() << '\n';
+               //std::cerr << "bathymetry.latitude_max(): " << bathymetry.latitude_max() << '\n';
+               //return bathymetry.latitude_min() + bathymetry.latitude_spacing() * (double)y;
+               return bathymetry.latitude_max() - bathymetry.latitude_spacing() * (double)y;
        }

        Vertex Geometry::get_earth_centered_vertex(double longitude, double latitude, double altitude, uint32_t id)
        {
+               // Test with:
+               // longitude, latitude, altitude: -122.503094 37.058946 -1020
+               // x, y, z:                       -2737902.2652    -4297133.60787    3822000.89563
+               //longitude = -122.503094;
+               //latitude = 37.058946;
+               //altitude = -1020;
+
+               //std::cerr << std::fixed;
+               //std::cerr << "longitude, latitude, altitude: " << longitude << ' ' << latitude << ' ' << altitude << '\n';
                double phi = to_radians(latitude);
                double theta = to_radians(longitude);

@@ -39,12 +55,41 @@ namespace stoqs
                double rho = EARTH_RADIUS_M + altitude;

                // this assumes z is up
-               float x = rho * cos_phi * cos_theta;
-               float y = rho * cos_phi * sin_theta;
-               float z = rho * sin_phi;
+               double x = rho * cos_phi * cos_theta;
+               double y = rho * cos_phi * sin_theta;
+               double z = rho * sin_phi;
+
+               //std::cerr << "x, y, z: " << x << ' ' << y << ' ' << z << '\n';
+
+               // Mimic https://github.com/GenericMappingTools/gmt/blob/be890649579be45e94269632786d08890a26dfea/src/gmt_map.c#L9094-L9108
+               // and   https://github.com/x3dom/x3dom/blob/3ace18318cd192e424546569932abfe3e1e2346a/src/nodes/Geospatial/GeoCoordinate.js#L380-L443
+               double F = 1.0 / WGS_84_INVERSE_FLATTENING;
+               double e_squared = F * ( 2.0 - F );
+
+               double sin_lon = sin(to_radians(longitude));
+               double cos_lon = cos(to_radians(longitude));
+               double sin_lat = sin(to_radians(latitude));
+               double cos_lat = cos(to_radians(latitude));
+
+               double N = WGS_84_SEMI_MAJOR_AXIS / sqrt (1.0 - e_squared * sin_lat * sin_lat);
+               double tmp = (N + altitude) * cos_lat;
+               x = tmp * cos_lon;
+               y = tmp * sin_lon;
+               z = (N * (1 - e_squared) + altitude) * sin_lat;
+
+               //std::cerr << "WGS-84 x, y, z: " << x << ' ' << y << ' ' << z << '\n';
+               // [vagrant@localhost build]$ ./grd-to-gltf Monterey25.grd -e 10 -b
+               // longitude, latitude, altitude: -122.503094 36.440848 -1020.000000
+               // x, y, z: -2753604.329500 -4321778.000725 3783720.828683
+               // WGS-84 x, y, z: -2759951.394247 -4331739.709638 3767050.208608
+               // [vagrant@localhost build]$ echo "-122.503094 36.440848 -1020.000000" | gmt mapproject -E
+               // -2759951.41997    -4331739.72409    3767050.17337
+               //exit(0);
+

                // gltf assumes y is up
-               return Vertex(x, z, y, id);
+               // With "return Vertex(x, z, y, id)" terrain ends up south of Australia
+               return Vertex(x, y, z, id);
        }
```
I left some of my debug print statements in (commented out).

**Comment by MBARIMike** [comment]
Created a .glb file with:
```
./grd-to-gltf Monterey25.grd -e 10 -b
cp Monterey25.glb Monterey25_e10_lat_rev.glb
```
Placed file on stoqs.mbari.org server and loaded a local stoqs database for testing with CTD data and other terrain models.

Here's a screen grab showing quite good agreement with the original Monterey_10x terrain:
Screen Shot 2021-12-15 at 12 35 25 PM

**Comment by MBARIMike** [comment]
With only the new .glb rendered we see all of the details of the bathymetry without the smoothing that exists in the 10x terrain:
Screen Shot 2021-12-15 at 12 41 14 PM
This is fantastic!

**Comment by stoqs** [comment]
Great work! Thanks for this contribution!

**Comment by ikehirzel** [comment]
Somes notes on building the gltf-generator on windows:

The windows machine will need few things to build the glTF generator:
- A valid C++11 compiler
- CMake
- [The NetCDF Library](https://www.unidata.ucar.edu/downloads/netcdf/)

The CMake build script should automatically find the C library dependency, but I believe the library was updated to use C++.
If there are issues with it being able to find dependencies, the script will likely need be manually updated to link it.

Other than that, it should build like a normal cmake project, producing gltf-generator.exe as the target applcation binary.

**Pull request merged**